### PR TITLE
Initial PEI memory buckets commit.  Integration with DXE will be adde…

### PR DESCRIPTION
**Pull request by kenlautner**
## Description

Added memory buckets to PEI for all memory types that persist into runtime. This is to reduce memory fragmentation for memory that is passed into the OS. Currently only EfiRuntimeServicesData is allocated in PEI.

MemoryBucketLib is added to keep track of PEI memory allocations of relevant types (currently only memory that persists into runtime) and consolidate their allocations to reduce fragmentation.

MemoryBucketLib keeps track of all the relevant information about the PEI memory buckets. It considers the newly added PeiMemoryBucket PCDs to decide what memory allocations to keep track of and possibly not be used at all. If all the PCDs remain at their default value of 0 PEI memory buckets will not be used and PEI memory allocation will function the same as it has before this change. Otherwise PEI memory buckets will be created for the memory types with non-zero PCD values.

These buckets are created in post-mem PEI and there is currently no support for pre-mem buckets.

- [x] Impacts functionality?
- [ ] Impacts security?
- [ ] Breaking change?
- [ ] Includes tests?
- [ ] Includes documentation?

## How This Was Tested

Currently there is an exception occurring on Physical and Virtual platforms.  Will continue to fine tune to get memory behaving correctly.

## Integration Instructions

N/A

### CODE REVIEW

**Comment by kenlautner** [comment]
Link to old PR: https://github.com/microsoft/mu_basecore/pull/167

File: MdeModulePkg/Include/Guid/MemoryTypeStatistics.h [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  EFI_PHYSICAL_ADDRESS    BaseAddress;
+  EFI_PHYSICAL_ADDRESS    MaximumAddress;
+  UINT64                  CurrentNumberOfPages;
+  UINT64                  NumberOfPages;
+  UINT32                  InformationIndex;
+  BOOLEAN                 Special;
+  BOOLEAN                 Runtime;
+} EFI_MEMORY_TYPE_STATISTICS;
+
+#endif
```
Comment (makubacki) [inline MdeModulePkg/Include/Guid/MemoryTypeStatistics.h:30]:
```suggestion
#endif

```

File: MdeModulePkg/Core/Pei/PeiMain.inf [hunk 2]
```diff
@@ -81,6 +83,8 @@
   gEdkiiMigratedFvInfoGuid                      ## SOMETIMES_PRODUCES     ## HOB
   gEfiFirmwarePerformanceGuid # MS_CHANGE_161871 - needed to build SEC perf HOB
   gEfiDelayedDispatchTableGuid   # MSCHANGE
+  gEfiMemoryTypeInformationGuid
+  gMemoryTypeStatisticsGuid
```
Comment (makubacki) [inline MdeModulePkg/Core/Pei/PeiMain.inf:87]:
```suggestion
  gEfiMemoryTypeInformationGuid   # MU_CHANGE
  gMemoryTypeStatisticsGuid       # MU_CHANGE
```

File: MdeModulePkg/Core/Pei/Memory/MemoryBuckets.c [hunk 3]
```diff
@@ -0,0 +1,470 @@
...
+  UINT64  AddressAdjustment;
+
+  AddressAdjustment = 0;
+
+  for (Index = 0; Index < EfiMaxMemoryType+1; Index++) {
+    PrivateData->PeiMemoryBuckets.RuntimeBuckets[Index].BaseAddress = 0;
+    PrivateData->PeiMemoryBuckets.RuntimeBuckets[Index].MaximumAddress = MAX_ALLOC_ADDRESS;
+    PrivateData->PeiMemoryBuckets.RuntimeBuckets[Index].CurrentNumberOfPages = 0;
+    PrivateData->PeiMemoryBuckets.RuntimeBuckets[Index].NumberOfPages = 0;
+    PrivateData->PeiMemoryBuckets.RuntimeBuckets[Index].InformationIndex = mMemoryTypes[Index];
```
Comment (makubacki) [inline MdeModulePkg/Core/Pei/Memory/MemoryBuckets.c:145]: Looks to me like this should just be `Index` based on usage I see elsewhere.

```suggestion
    PrivateData->PeiMemoryBuckets.RuntimeBuckets[Index].InformationIndex = Index;

```

**Comment by makubacki** [comment]
Closed in favor of https://github.com/microsoft/mu_basecore/pull/267

**Pull request closed without merging**